presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,245 @@
+        if (dataStream == null && presentStream != null) {
+            streamPosition = readAllNulls(positions, positionCount);
+        }
+        else {
+            for (int i = 0; i < positionCount; i++) {
+                int position = positions[i];
+                if (position > streamPosition) {
+                    skip(position - streamPosition);
+                    streamPosition = position;
+                }
+
+                if (presentStream != null && !presentStream.nextBit()) {
+                    if (nullsAllowed) {
+                        if (outputRequired) {
+                            nulls[outputPositionCount] = true;
+                        }
+                        if (filter != null) {
+                            outputPositions[outputPositionCount] = position;
+                        }
+                        outputPositionCount++;
+                    }
+                }
+                else {
+                    long value = dataStream.next();
+                    if (filter == null || filter.testLong(value)) {
+                        if (outputRequired) {
+                            values[outputPositionCount] = value;
+                            if (nullsAllowed && presentStream != null) {
+                                nulls[outputPositionCount] = false;
+                            }
+                        }
+                        if (filter != null) {
+                            outputPositions[outputPositionCount] = position;
+                        }
+                        outputPositionCount++;
+                    }
+                }
+                streamPosition++;
+            }
+        }
+
+        readOffset = offset + streamPosition;
+
+        return outputPositionCount;
+    }
+
+    private int readAllNulls(int[] positions, int positionCount)
+            throws IOException
+    {
+        presentStream.skip(positions[positionCount - 1]);
+
+        if (nullsAllowed) {
+            outputPositionCount = positionCount;
+            allNulls = true;
+        }
+        else {
+            outputPositionCount = 0;
+        }
+
+        return positions[positionCount - 1] + 1;
+    }
+
+    private void skip(int items)
+            throws IOException
+    {
+        if (presentStream != null) {
+            int dataToSkip = presentStream.countBitsSet(items);
+            dataStream.skip(dataToSkip);
+        }
+        else {
+            dataStream.skip(items);
+        }
+    }
+
+    private void openRowGroup()
+            throws IOException
+    {
+        presentStream = presentStreamSource.openStream();
+        dataStream = dataStreamSource.openStream();
+
+        rowGroupOpen = true;
+    }
+
+    @Override
+    public Block getBlock(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
```

Comment:
Even `outputRequired` implies `outputType` is not null, maybe it's still good to have `checkState(outputType != null, ...)`

Reply:
I don't think an extra check is necessary. These variables are managed locally and therefore it is safe to assume what their values are. If outputRequired is true, but outputType is null, then there is a bug in this class.